pkg/services/alerting/eval_context.go

```diff
@@ -188,11 +190,15 @@ func getNewStateInternal(c *EvalContext) models.AlertStateType {
 // evaluateNotificationTemplateFields will treat the alert evaluation rule's name and message fields as
 // templates, and evaluate the templates using data from the alert evaluation's tags
 func (c *EvalContext) evaluateNotificationTemplateFields() error {
+	matches := c.EvalMatches
 	if len(c.EvalMatches) < 1 {
-		return nil
+		if len(c.AllSeries) < 1 {
+			return nil
+		}
+		matches = c.AllSeries
```

Comment:
After our offline discussion, I don't think this is correct.

What this section is doing is: Use all the matches if you don't have any alerting matches (given `EvalMatches` only contain those values that would be alerting).

Let's think about a scenario for a second: Let's say you have an alert that observes 5 series

1. All 5 series go into alert, you get an alert that says: "ALERTING s1,s2,s3,s4,s5"
2. On the next iteration 2 them resolve, what you need is two notifications: "ALERTING s1,s2,s3" and "OK s4,s5"
3. On the next iteration all of them resolve "OK s1,s2,s3" 

How does this map to this understanding? In my eyes (with my limited knowledge about how templating works in legacy) what you need is to compute the difference between `EvalMatches` and `AllMatches` and assume those are the non-firing alerts.

Reply:
Once we figure the above, I'd recommend you scope this out to a function so that you can do something like:

```golang
func (c *EvalContext) GetTemplateMatches() []*EvalMatch {
	// EvalMatches represent alerting values, if we have any this means this is a firing alert and we should use those the data for the template.
	if len(c.EvalMatches) > 0 {
		return c.EvalMatches
	}

	// If we don't have any alerting values, use all of the values to template the alert.
	return c.AllMatches
}
```

Reply:
Making that diff between `EvalMatches` and `AllMatches` would result in a set of series that didn't violate the constraint for the alert to fire. In that case it would be possible to send an extra `OK` notification.

For item 3 we would need the previous evaluation data. Without that we can't know what which series were firing and got resolved.

Reply:
With this PR as it is right now, what would happen at step 2? Nothing? Or will it say "OK s1,s2,s3, s4, s5" despite s1, s2, and s3 still being alerting?

Reply:
It would say "ALERTING s1,s2,s3", but we have nothing in place to send an extra OK notification for the series that are no longer firing.
I reproduced this scenario. This is what happens if the notifier is configured to send reminders:

![Screen Shot 2022-04-13 at 09 06 25](https://user-images.githubusercontent.com/41638679/163176304-93f261bf-7b9e-4c3e-9f33-9f4d1d752f17.png)

Two series were causing the alert to fire, then just one, and then all series were OK.
The alert rule name is `Prueba legacy alert: ${instance}`.